library/alloc/src/vec/mod.rs

```diff
@@ -2309,31 +2308,39 @@ impl<T, A: Allocator> Vec<T, A> {
         // In cases when predicate and `drop` never panick, it will be optimized out.
         struct BackshiftOnDrop<'a, T, A: Allocator> {
             v: &'a mut Vec<T, A>,
-            processed_len: usize,
-            deleted_cnt: usize,
+            read_index: usize,
+            write_index: usize,
             original_len: usize,
         }
 
         impl<T, A: Allocator> Drop for BackshiftOnDrop<'_, T, A> {
             fn drop(&mut self) {
-                if self.deleted_cnt > 0 {
+                if self.read_index < self.original_len {
+                    self.write_index = self.write_index.min(self.read_index);
+                    let remaining = self.original_len - self.read_index;
                     // SAFETY: Trailing unchecked items must be valid since we never touch them.
                     unsafe {
                         ptr::copy(
-                            self.v.as_ptr().add(self.processed_len),
-                            self.v.as_mut_ptr().add(self.processed_len - self.deleted_cnt),
-                            self.original_len - self.processed_len,
+                            self.v.as_ptr().add(self.read_index),
+                            self.v.as_mut_ptr().add(self.write_index),
+                            remaining,
                         );
                     }
+                    // SAFETY: After filling holes, all items are in contiguous memory.
+                    unsafe {
+                        self.v.set_len(self.write_index + remaining);
+                    }
+                    return;
                 }
-                // SAFETY: After filling holes, all items are in contiguous memory.
+                // SAFETY: no panic happened, so all items are in contiguous memory.
                 unsafe {
-                    self.v.set_len(self.original_len - self.deleted_cnt);
+                    self.v.set_len(self.write_index.min(self.read_index));
                 }
             }
         }
 
-        let mut g = BackshiftOnDrop { v: self, processed_len: 0, deleted_cnt: 0, original_len };
+        let mut g =
+            BackshiftOnDrop { v: self, read_index: 0, write_index: usize::MAX, original_len };
```

Comment:
Instead of using `usize::MAX` and the `min` call above, I think it'd be better to use a `loop` for the first phase and return from that if `read_index == original_len`. That way, you can properly initialise `write_index` upon exiting the `loop`, when you know that there are items to delete.